S4TF_Tutorial_2.ipynb

```diff
@@ -0,0 +1,875 @@
+        "*   We trained it with 3500 examples (7 pairs, over 500 epochs).\n",
+        "\n",
+        "Our model tuned the variables (weights) in the Dense layer until it was able to return the correct Fahrenheit value for any Celsius value. (Remember, none of these degree Celsius temperatures were part of our training data.)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "ndGNLPYjzflz",
+        "colab_type": "text"
+      },
+      "source": [
+        "## Looking at the layer weights\n",
+        "\n",
+        "Finally, let's print the internal variables of the Dense layer. "
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "e3CYb7sqxr9i",
+        "colab_type": "code",
+        "outputId": "e18d1aff-2896-48e8-f6aa-7a9c3e30a840",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 51
+        }
+      },
+      "source": [
+        "print(model.l0.weight)\n",
+        "print(model.l0.bias)"
+      ],
+      "execution_count": 11,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "[[1.8245275]]\r\n",
+            "[28.855717]\r\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "YwPV0Orhzh8h",
+        "colab_type": "text"
+      },
+      "source": [
+        "The first variable is close to ~1.8 and the second to ~32. These values (1.8 and 32) are the actual variables in the real conversion formula.\n",
+        "\n",
+        "This is really close to the values in the conversion formula. We'll explain this in an upcoming video where we show how a Dense layer works, but for a single neuron with a single input and a single output, the internal math looks the same as [the equation for a line](https://en.wikipedia.org/wiki/Linear_equation#Slope%E2%80%93intercept_form), $y = mx + b$, which has the same form as the conversion equation, $f = 1.8c + 32$.\n",
+        "\n",
+        "Since the form is the same, the variables should converge on the standard values of 1.8 and 32, which is exactly what happened.\n",
+        "\n",
+        "With additional neurons, additional inputs, and additional outputs, the formula becomes much more complex, but the idea is the same. \n",
+        "\n",
+        "### A little experiment\n",
+        "\n",
+        "Just for fun, what if we created more Dense layers with different units, which therefore also has more variables?"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "Yz7ATykXxtSA",
+        "colab_type": "code",
+        "outputId": "60ec62a4-2fd2-4d18-a79f-75ae3e3baff8",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 1004
+        }
+      },
+      "source": [
+        "let hiddenSize: Int = 4\n",
+        "struct c2fConvExp: Layer {\n",
```

Comment:
Type names should start with an uppercase letter. 

Also, what does "Exp" mean? The Swift API Design Guidelines actually discourage abbreviations. How about renaming `c2fConv` to `CelsiusToFahrenheit`, and renaming `c2fConvExp` to `CelsiusToFahrenheit2`?

Reply:
This was a part of an experiment for the learners, so I named it as Exp

Reply:
Understood. However, since the entire section is titled "A little experiment", adding "exp" to each variable feels not adding much value but just redundancy. Also, because Swift generally doesn't encourage the use of abbreviations and because "exp" has many possible meanings, names ending with "Exp" can be very confusing.